nixos/modules/security/auditd.nix

```diff
@@ -4,16 +4,260 @@
+    };
+
+    plugins = lib.mkOption {
+      type = lib.types.attrsOf plugin_conf;
+      default = { };
+      defaultText = lib.literalExpression ''
+        {
+          af_unix = {
+            path = lib.getExe' pkgs.audit "audisp-af_unix";
+            args = [
+              "0640"
+              "/var/run/audispd_events"
+              "string"
+            ];
+            format = "binary";
+          };
+          remote = {
+            path = lib.getExe' pkgs.audit "audisp-remote";
+            config = { };
+          };
+          filter = {
+            path = lib.getExe' pkgs.audit "audisp-filter";
+            args = [
+              "allowlist"
+              "/etc/audit/audisp-filter.conf"
+              (lib.getExe' pkgs.audit "audisp-syslog")
+              "LOG_USER"
+              "LOG_INFO"
+              "interpret"
+            ];
+            config = { };
+          };
+          syslog = {
+            path = lib.getExe' pkgs.audit "audisp-syslog";
+            args = [ "LOG_INFO" ];
+          };
+        }
+      '';
+      description = "Plugin definitions to register with auditd";
+    };
+  };
+
+  config = lib.mkIf cfg.enable {
+    assertions = [
+      {
+        assertion =
+          let
+            cfg' = cfg.config;
+          in
+          (
+            (lib.isInt cfg'.space_left && lib.isInt cfg'.admin_space_left)
+            -> cfg'.space_left > cfg'.admin_space_left
+          )
+          && (
+            let
+              get_percent = s: lib.toInt (lib.strings.removeSuffix "%" s);
+            in
+            (lib.isString cfg'.space_left && lib.isString cfg'.admin_space_left)
+            -> (get_percent cfg'.space_left) > (get_percent cfg'.admin_space_left)
+          );
+        message = "`security.auditd.config.space_left` must be larger than `security.auditd.config.admin_space_left`";
+      }
+    ];
 
-  config = lib.mkIf config.security.auditd.enable {
     # Starting auditd should also enable loading the audit rules..
     security.audit.enable = lib.mkDefault true;
 
     environment.systemPackages = [ pkgs.audit ];
 
+    # setting this to anything other than /etc/audit/plugins.d will break, so we pin it here
+    security.auditd.config.plugin_dir = "/etc/audit/plugins.d";
+
+    environment.etc = {
+      "audit/auditd.conf".text = prepareConfigText cfg.config;
+    }
+    // (lib.mapAttrs' (
+      plugin_name: plugin_definition_config_value:
+      lib.nameValuePair "audit/plugins.d/${plugin_name}.conf" {
+        text = prepareConfigText (lib.removeAttrs plugin_definition_config_value [ "config" ]);
+      }
+    ) cfg.plugins)
+    // (lib.mapAttrs' (
+      plugin_name: plugin_specific_config_value:
+      lib.nameValuePair "audit/audisp-${plugin_name}.conf" {
+        text = prepareConfigText plugin_specific_config_value.config;
+      }
+    ) (lib.filterAttrs (_: v: v.config != null) cfg.plugins));
+
+    security.auditd.plugins = {
```

Comment:
Be aware of overriding parts of this config. The way config is merged based on it's priority is no t immediately obvious. See https://github.com/NixOS/nixpkgs/pull/285114

There is no action you can take right now but when this module is used, this might come up in the future.